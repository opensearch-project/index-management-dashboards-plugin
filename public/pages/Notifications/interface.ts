/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ActionTypeMapName } from "./constant";

type ValueOf<T> = T[keyof T];
export interface ILronConfig {
  lron_condition: {
    success?: boolean;
    failure?: boolean;
  };
  task_id?: string;
  action_name?: ValueOf<typeof ActionTypeMapName>;
  channels: Array<{
    id: string;
  }>;
}

export type ILronPlainConfig = ILronConfig["lron_condition"] &
  Omit<ILronConfig, "lron_condition"> & {
    title: string;
    index: number;
  };

export interface FieldState {
  dataSource: ILronPlainConfig[];
  allConfig: ILronPlainConfig;
}
