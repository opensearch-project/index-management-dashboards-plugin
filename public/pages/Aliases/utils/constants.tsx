/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SortDirection } from "../../../utils/constants";
import { IAlias } from "../interface";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: "0",
  size: "20",
  search: "",
  sortField: "alias" as keyof IAlias,
  sortDirection: SortDirection.DESC,
  status: "",
  dataSourceId: "",
};
