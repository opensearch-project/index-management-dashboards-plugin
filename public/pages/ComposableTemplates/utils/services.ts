/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonService } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";

export const getAllUsedComponents = async ({ commonService }: { commonService: CommonService }) => {
  const allTemplatesResponse = await commonService.apiCaller<{
    index_templates?: Array<{
      name: string;
      index_template: TemplateItemRemote;
    }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: "/_index_template/*",
    },
  });

  if (allTemplatesResponse && allTemplatesResponse.ok) {
    return allTemplatesResponse.response.index_templates || [];
  }

  return [];
};
