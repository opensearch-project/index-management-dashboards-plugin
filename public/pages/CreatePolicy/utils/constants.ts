/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const DEFAULT_POLICY = JSON.stringify(
  {
    policy: {
      description: "A simple default policy that changes the replica count between hot and cold states.",
      default_state: "example_hot_state",
      states: [
        {
          name: "example_hot_state",
          actions: [{ replica_count: { number_of_replicas: 5 } }],
          transitions: [{ state_name: "example_cold_state", conditions: { min_index_age: "30d" } }],
        },
        {
          name: "example_cold_state",
          actions: [{ replica_count: { number_of_replicas: 2 } }],
          transitions: [],
        },
      ],
      ism_template: {
        index_patterns: ["example-index-*"],
      },
    },
  },
  null,
  4
);
