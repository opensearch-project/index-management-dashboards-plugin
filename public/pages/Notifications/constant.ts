export enum ActionType {
  REINDEX = "REINDEX",
  RESIZE = "RESIZE",
  FORCEMERGE = "FORCEMERGE",
  OPEN = "OPEN",
}

export enum OperationType {
  REINDEX = "REINDEX",
  FORCEMERGE = "FORCEMERGE",
  OPEN = "OPEN",
  SHRINK = "SHRINK",
  CLONE = "CLONE",
  SPLIT = "SPLIT",
}

export const ActionTypeMapTitle = {
  [ActionType.REINDEX]: "reindex",
  [ActionType.RESIZE]: "shrink, split, clone",
  [ActionType.FORCEMERGE]: "force merge",
  [ActionType.OPEN]: "open",
};

export const ActionTypeMapDescription = {
  [ActionType.REINDEX]: "Notify when any reindex operation has failed or completed.",
  [ActionType.RESIZE]: "Notify when any shrink, split, or clone operation has failed or completed.",
  [ActionType.FORCEMERGE]: "Notify when any force merge operation has failed or completed.",
  [ActionType.OPEN]: "Notify when any open operation has failed or completed.",
};

export const OperationTypeMapTitle = {
  ...ActionTypeMapTitle,
  [OperationType.SHRINK]: "shrink",
  [OperationType.SPLIT]: "split",
  [OperationType.CLONE]: "clone",
};

export const ActionTypeMapName = {
  [ActionType.REINDEX]: "indices:data/write/reindex",
  [ActionType.RESIZE]: "indices:admin/resize",
  [ActionType.FORCEMERGE]: "indices:admin/forcemerge",
  [ActionType.OPEN]: "indices:admin/open",
} as const;

export const getKeyByValue = <T extends {}, Key extends keyof T>(obj: T, value: T[Key]) => {
  if (!obj) {
    return undefined;
  }
  return Object.keys(obj).find((key) => obj[key as Key] === value);
};

export enum FieldEnum {
  channels = "channels",
  failure = "failure",
  success = "success",
}

export const FieldMapLabel = {
  [FieldEnum.channels]: "Notification channels",
};

export const LABEL_FOR_CONDITION = "Notify when operation";
export const VALIDATE_ERROR_FOR_CHANNELS = "One or more channels is required.";
