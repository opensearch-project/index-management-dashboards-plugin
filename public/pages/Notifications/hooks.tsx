/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { CommonService } from "../../services";
import { ActionType, ActionTypeMapName, ActionTypeMapTitle } from "./constant";
import { ILronConfig, ILronPlainConfig } from "./interface";

export const transformConfigListToPlainList = (config: ILronConfig[]): ILronPlainConfig[] => {
  return Object.entries(ActionTypeMapName).map(([actionType, action_name], index) => {
    const findItem =
      config.find((c) => c.action_name === action_name) ||
      ({
        lron_condition: {
          failure: false,
          success: false,
        },
        action_name,
        channels: [],
      } as ILronConfig);
    const { lron_condition, ...rest } = findItem;
    return {
      ...lron_condition,
      ...rest,
      index,
      title: actionType ? ActionTypeMapTitle[actionType as ActionType] : "",
    };
  });
};

export const transformPlainListToConfigList = (config: ILronPlainConfig[]): ILronConfig[] => {
  return Object.entries(ActionTypeMapName).map(([actionType, action_name], currentIndex) => {
    const findItem =
      config.find((c) => c.action_name === action_name) ||
      ({
        failure: false,
        success: false,
        action_name,
        title: ActionTypeMapTitle[actionType as ActionType],
        channels: [],
        index: currentIndex,
      } as ILronPlainConfig);
    const { failure, success, index, title, ...rest } = findItem;
    return {
      ...rest,
      lron_condition: {
        failure,
        success,
      },
    } as ILronConfig;
  });
};

export const getComputedResultFromPlainList = (
  plainConfigs: ILronPlainConfig[]
): {
  useDifferentSettings: boolean;
} => {
  return {
    useDifferentSettings: !Object.values(ActionTypeMapName).every((c) => {
      const findItem = plainConfigs.find((p) => p.action_name === c) || ({} as ILronPlainConfig);
      return findItem.success === plainConfigs[0].failure && findItem.success === plainConfigs[0].success;
    }),
  };
};

export const getDiffableMapFromPlainList = (plainConfigs: ILronPlainConfig[]): Record<string, ILronPlainConfig> => {
  return plainConfigs.reduce(
    (total, current) => ({
      ...total,
      [current.action_name as string]: current,
    }),
    {}
  );
};

export const submitNotifications = async (props: { commonService: CommonService; plainConfigsPayload: ILronPlainConfig[] }) => {
  return Promise.all(
    transformPlainListToConfigList(props.plainConfigsPayload).map((item) => {
      const { action_name } = item;
      return props.commonService.apiCaller<unknown>({
        endpoint: "transport.request",
        data: {
          method: "PUT",
          path: `/_plugins/_im/lron/${encodeURIComponent(`LRON:${action_name}`)}`,
          body: {
            lron_config: item,
          },
        },
      });
    })
  ).then((allResults) => {
    return {
      ok: allResults.every((item) => item.ok),
      body: allResults.filter((item) => item.body).map((item) => item.body),
      error: allResults
        .filter((item) => !item.ok)
        .map((item) => item.error)
        .join(", "),
    };
  });
};

export const getNotifications = async (props: { commonService: CommonService }) => {
  return props.commonService
    .apiCaller<{
      lron_configs: Array<{
        lron_config: ILronConfig;
      }>;
      total_number: number;
    }>({
      endpoint: "transport.request",
      data: {
        method: "GET",
        path: "/_plugins/_im/lron",
      },
    })
    .then((res) => {
      if (res.ok) {
        return res;
      } else if (res.body?.error?.type === "index_not_found_exception") {
        return {
          ok: true,
          response: {
            lron_configs: [],
            total_number: 0,
          },
          error: "",
          body: {},
        };
      } else {
        return res;
      }
    });
};
