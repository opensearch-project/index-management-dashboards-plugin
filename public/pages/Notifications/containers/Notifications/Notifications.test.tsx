/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { MemoryRouter as Router, Route, RouteComponentProps, Switch } from "react-router-dom";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Notifications from "./Notifications";
import { ServicesContext } from "../../../../services";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import { ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";
import { getApplication, getNavigationUI, getUISettings } from "../../../../services/Services";

jest.mock("../../../../services/Services", () => ({
  ...jest.requireActual("../../../../services/Services"),
  getUISettings: jest.fn(),
  getApplication: jest.fn(),
  getNavigationUI: jest.fn(),
}));

beforeEach(() => {
  (getUISettings as jest.Mock).mockReturnValue({
    get: jest.fn().mockReturnValue(false), // or false, depending on your test case
  });
  (getApplication as jest.Mock).mockReturnValue({});

  (getNavigationUI as jest.Mock).mockReturnValue({});
});

function renderNotificationsWithRouter(initialEntries = [ROUTES.NOTIFICATIONS] as string[]) {
  return {
    ...render(
      <Router initialEntries={initialEntries}>
        <CoreServicesContext.Provider value={coreServicesMock}>
          <ServicesContext.Provider value={browserServicesMock}>
            <Switch>
              <Route path={ROUTES.NOTIFICATIONS} render={(props: RouteComponentProps) => <Notifications {...props} />} />
            </Switch>
          </ServicesContext.Provider>
        </CoreServicesContext.Provider>
      </Router>
    ),
  };
}

describe("<Notifications /> spec", () => {
  beforeEach(() => {
    browserServicesMock.commonService.apiCaller = jest.fn(
      async (payload): Promise<any> => {
        switch (payload.endpoint) {
          case "transport.request": {
            if (payload.data?.path?.startsWith("/_plugins/_im/lron")) {
              return {
                ok: true,
                response: {
                  lron_configs: [],
                  total_number: 0,
                },
              };
            } else {
              return {
                ok: true,
                response: {},
              };
            }
          }
        }
        return {
          ok: true,
          response: {},
        };
      }
    );
    browserServicesMock.notificationService.getChannels = jest.fn(
      async (): Promise<any> => {
        return {
          ok: true,
          response: {
            start_index: 0,
            total_hits: 1,
            total_hit_relation: "eq",
            channel_list: [
              {
                config_id: "1",
                name: "channel1",
                description: "2",
                config_type: "chime",
                is_enabled: true,
              },
            ],
          },
        };
      }
    );
  });

  it("renders", async () => {
    const { container, findByText } = renderNotificationsWithRouter([ROUTES.NOTIFICATIONS]);
    await findByText("Notification settings");
    expect(container).toMatchSnapshot();
    await findByText("reindex");
  });

  it("Update notification settings", async () => {
    const { findByText, getByTestId, getByText, queryByTestId } = renderNotificationsWithRouter([ROUTES.NOTIFICATIONS]);
    await findByText("reindex");
    await userEvent.click(getByTestId("dataSource.0.failure"));
    await findByText("Save");
    await userEvent.click(getByText("Save"));
    await findByText("Address the following error(s) in the form");
    await userEvent.type(
      getByTestId("dataSource.0.channels")?.querySelector('[data-test-subj="comboBoxSearchInput"]') as Element,
      "channel1{enter}"
    );
    await userEvent.click(getByTestId("dataSource.0.failure"));
    await waitFor(
      () => {
        expect(queryByTestId("dataSource.0.channels")).toBeNull();
      },
      {
        timeout: 3000,
      }
    );
    await userEvent.click(getByTestId("dataSource.0.failure"));
    await waitFor(() => {
      expect(queryByTestId("dataSource.0.channels")).not.toBeNull();
    });

    await userEvent.click(getByText("Save"));
    await waitFor(() => {
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledWith(
        "Notifications settings for index operations have been successfully updated."
      );
    });
  });

  it("View without permission", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn(
      async (payload): Promise<any> => {
        switch (payload.endpoint) {
          case "transport.request": {
            if (payload.data?.path?.startsWith("/_plugins/_im/lron")) {
              return {
                ok: false,
                body: {
                  status: 403,
                },
              };
            } else {
              return {
                ok: true,
                response: {},
              };
            }
          }
        }
        return {
          ok: true,
          response: {},
        };
      }
    );
    const { findByText, container } = renderNotificationsWithRouter([ROUTES.NOTIFICATIONS]);
    await findByText("Error loading Notification settings");
    await waitFor(() => {
      expect(container).toMatchSnapshot();
      expect(coreServicesMock.notifications.toasts.addDanger).toBeCalledWith({
        title: "You do not have permissions to view notification settings",
        text: "Contact your administrator to request permissions.",
      });
    });
  });

  it("Update without permission", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn(
      async (payload): Promise<any> => {
        switch (payload.endpoint) {
          case "transport.request": {
            if (payload.data?.path?.includes("?dry_run=true")) {
              return {
                ok: false,
              };
            } else if (payload.data?.path?.startsWith("/_plugins/_im/lron")) {
              return {
                ok: true,
                response: {
                  lron_configs: [],
                  total_number: 0,
                },
              };
            } else {
              return {
                ok: true,
                response: {},
              };
            }
          }
        }
        return {
          ok: true,
          response: {},
        };
      }
    );
    const { findByText, getByTestId, getByText, queryByText } = renderNotificationsWithRouter([ROUTES.NOTIFICATIONS]);
    await findByText("reindex");
    await userEvent.click(getByTestId("dataSource.0.failure"));
    await findByText("Save");
    await userEvent.click(getByText("Save"));
    await waitFor(() =>
      expect(coreServicesMock.notifications.toasts.addDanger).toBeCalledWith({
        title: "You do not have permissions to update notification settings",
        text: "Contact your administrator to request permissions.",
      })
    );
    await userEvent.click(getByText("Cancel"));
    await waitFor(() => {
      expect(queryByText("Cancel")).toBeNull();
    });
  });

  it("Update with auto populate channels", async () => {
    const { container, findByText, getByTestId, findByTestId } = renderNotificationsWithRouter([ROUTES.NOTIFICATIONS]);
    await findByText("reindex");
    await userEvent.click(getByTestId("dataSource.0.failure"));
    await waitFor(() => {
      expect(getByTestId("dataSource.0.channels").querySelector(".euiLoadingSpinner")).toBeNull();
    });
    await userEvent.type(
      getByTestId("dataSource.0.channels")?.querySelector('[data-test-subj="comboBoxSearchInput"]') as Element,
      "channel1{enter}"
    );
    await userEvent.click(getByTestId("dataSource.1.failure"));
    await findByTestId("dataSource.1.channels");
    await waitFor(() => expect(container.querySelector(`[data-test-subj="dataSource.1.channels"] [title="channel1"]`)).toBeInTheDocument());
  });
});
