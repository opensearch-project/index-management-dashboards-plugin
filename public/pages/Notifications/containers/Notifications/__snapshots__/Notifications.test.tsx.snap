// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> spec View without permission 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem"
    >
      <div
        class="euiText euiText--small"
      >
        <h1>
          Notification settings
        </h1>
      </div>
      <div
        class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            style="padding-top: 0px; padding-bottom: 4px;"
          >
            Configure the default notification settings on index operation statuses, such as failed or completed. You can configure additional notification settings while performing an index operation.
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <a
        class="euiButton euiButton--primary euiButton--small"
        href="notifications-dashboards#/channels"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="euiButtonContent euiButtonContent--iconRight euiButton__content"
        >
          EuiIconMock
          <span
            class="euiButton__text"
          >
            Manage channels
          </span>
        </span>
      </a>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  >
    <div
      class="euiEmptyPrompt"
    >
      EuiIconMock
      <div
        class="euiSpacer euiSpacer--m"
      />
      <h2
        class="euiTitle euiTitle--medium"
      >
        Error loading Notification settings
      </h2>
      <span
        class="euiTextColor euiTextColor--subdued"
      >
        <div
          class="euiSpacer euiSpacer--m"
        />
        <div
          class="euiText euiText--medium"
        >
          <p>
            You do not have permissions to view Notification settings. Contact your administrator to request permissions.
          </p>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`<Notifications /> spec renders 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem"
    >
      <div
        class="euiText euiText--small"
      >
        <h1>
          Notification settings
        </h1>
      </div>
      <div
        class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            style="padding-top: 0px; padding-bottom: 4px;"
          >
            Configure the default notification settings on index operation statuses, such as failed or completed. You can configure additional notification settings while performing an index operation.
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <a
        class="euiButton euiButton--primary euiButton--small"
        href="notifications-dashboards#/channels"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="euiButtonContent euiButtonContent--iconRight euiButton__content"
        >
          EuiIconMock
          <span
            class="euiButton__text"
          >
            Manage channels
          </span>
        </span>
      </a>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  >
    <h2
      class="euiTitle euiTitle--small"
    >
      Defaults for index operations
    </h2>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiDescribedFormGroup euiDescribedFormGroup--fullWidth"
      role="group"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
          style="max-width: 700px;"
        >
          <div
            class="euiTitle euiTitle--xsmall euiDescribedFormGroup__title"
          >
            <div
              class="euiText euiText--small"
            >
              <h3
                class="ISM-notifications-first-letter-uppercase"
              >
                reindex
              </h3>
            </div>
          </div>
          <div
            class="euiText euiText--small euiDescribedFormGroup__description"
          >
            <div
              class="euiTextColor euiTextColor--subdued"
            >
              Notify when any reindex operation has failed or completed.
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiDescribedFormGroup__fields euiDescribedFormGroup__fieldPadding--xsmall"
        >
          <div
            class="euiFormRow euiFormRow--compressed"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormRow euiFormRow--compressed"
                id="some_html_id-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="some_html_id"
                  >
                    Notify when operation
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.0.failure"
                            id="builtInCheckBoxId-0"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-0"
                          >
                            Has failed
                          </label>
                        </div>
                      </span>
                    </div>
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.0.success"
                            id="builtInCheckBoxId-1"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-1"
                          >
                            Has completed
                          </label>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiDescribedFormGroup euiDescribedFormGroup--fullWidth"
      role="group"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
          style="max-width: 700px;"
        >
          <div
            class="euiTitle euiTitle--xsmall euiDescribedFormGroup__title"
          >
            <div
              class="euiText euiText--small"
            >
              <h3
                class="ISM-notifications-first-letter-uppercase"
              >
                shrink, split, clone
              </h3>
            </div>
          </div>
          <div
            class="euiText euiText--small euiDescribedFormGroup__description"
          >
            <div
              class="euiTextColor euiTextColor--subdued"
            >
              Notify when any shrink, split, or clone operation has failed or completed.
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiDescribedFormGroup__fields euiDescribedFormGroup__fieldPadding--xsmall"
        >
          <div
            class="euiFormRow euiFormRow--compressed"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormRow euiFormRow--compressed"
                id="some_html_id-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="some_html_id"
                  >
                    Notify when operation
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.1.failure"
                            id="builtInCheckBoxId-2"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-2"
                          >
                            Has failed
                          </label>
                        </div>
                      </span>
                    </div>
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.1.success"
                            id="builtInCheckBoxId-3"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-3"
                          >
                            Has completed
                          </label>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiDescribedFormGroup euiDescribedFormGroup--fullWidth"
      role="group"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
          style="max-width: 700px;"
        >
          <div
            class="euiTitle euiTitle--xsmall euiDescribedFormGroup__title"
          >
            <div
              class="euiText euiText--small"
            >
              <h3
                class="ISM-notifications-first-letter-uppercase"
              >
                force merge
              </h3>
            </div>
          </div>
          <div
            class="euiText euiText--small euiDescribedFormGroup__description"
          >
            <div
              class="euiTextColor euiTextColor--subdued"
            >
              Notify when any force merge operation has failed or completed.
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiDescribedFormGroup__fields euiDescribedFormGroup__fieldPadding--xsmall"
        >
          <div
            class="euiFormRow euiFormRow--compressed"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormRow euiFormRow--compressed"
                id="some_html_id-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="some_html_id"
                  >
                    Notify when operation
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.2.failure"
                            id="builtInCheckBoxId-4"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-4"
                          >
                            Has failed
                          </label>
                        </div>
                      </span>
                    </div>
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.2.success"
                            id="builtInCheckBoxId-5"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-5"
                          >
                            Has completed
                          </label>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiDescribedFormGroup euiDescribedFormGroup--fullWidth"
      role="group"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
          style="max-width: 700px;"
        >
          <div
            class="euiTitle euiTitle--xsmall euiDescribedFormGroup__title"
          >
            <div
              class="euiText euiText--small"
            >
              <h3
                class="ISM-notifications-first-letter-uppercase"
              >
                open
              </h3>
            </div>
          </div>
          <div
            class="euiText euiText--small euiDescribedFormGroup__description"
          >
            <div
              class="euiTextColor euiTextColor--subdued"
            >
              Notify when any open operation has failed or completed.
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiDescribedFormGroup__fields euiDescribedFormGroup__fieldPadding--xsmall"
        >
          <div
            class="euiFormRow euiFormRow--compressed"
            id="some_html_id-row"
          >
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                class="euiFormRow euiFormRow--compressed"
                id="some_html_id-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="some_html_id"
                  >
                    Notify when operation
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
                  >
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.3.failure"
                            id="builtInCheckBoxId-6"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-6"
                          >
                            Has failed
                          </label>
                        </div>
                      </span>
                    </div>
                    <div
                      class="euiFlexItem"
                    >
                      <span
                        class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                      >
                        <div
                          class="euiCheckbox euiCheckbox--compressed"
                        >
                          <input
                            class="euiCheckbox__input"
                            data-test-subj="dataSource.3.success"
                            id="builtInCheckBoxId-7"
                            type="checkbox"
                          />
                          <div
                            class="euiCheckbox__square"
                          />
                          <label
                            class="euiCheckbox__label"
                            for="builtInCheckBoxId-7"
                          >
                            Has completed
                          </label>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
